### PR TITLE
[MXNET-261]Update MKLDNN & Add CPP Test

### PR DESCRIPTION
**Pull request by xinyu-intel**
## Description ##
This pr aims to fix bugs in #8712 by update MKLDNN to the newest. CPP tests are added to monitor data format change of  MKL-DNN [MXNET-98](https://issues.apache.org/jira/browse/MXNET-98)

@pengzhao-intel

### Essentials ###
Please feel free to remove inapplicable items for your PR.
- [x] The PR title starts with [MXNET-$JIRA_ID], where $JIRA_ID refers to the relevant [JIRA issue](https://issues.apache.org/jira/projects/MXNET/issues) created (except PRs with tiny changes)
- [x] Changes are complete (i.e. I finished coding on this PR)
- [x] All changes have test coverage:
- Unit tests are added for small changes to verify correctness (e.g. adding a new operator)
- Nightly tests are added for complicated/long-running ones (e.g. changing distributed kvstore)
- Build tests will be added for build configuration changes (e.g. adding a new build option with NCCL)
- [x] Code is well-documented: 
- For user-facing API changes, API doc string has been updated. 
- For new C++ functions in header files, their functionalities and arguments are documented. 
- For new examples, README.md is added to explain the what the example does, the source of the dataset, expected performance on test set and reference to the original paper if applicable
- Check the API doc at http://mxnet-ci-doc.s3-accelerate.dualstack.amazonaws.com/PR-$PR_ID/$BUILD_ID/index.html
- [x] To the my best knowledge, examples are either not affected by this change, or have been fixed to be compatible with this change

### Changes ###
- [x] Update MKLDNN
- [x] Add cpp tests

### CODE REVIEW

**Comment by zheng-da** [comment]
Should we update MKLDNN to the latest commit in its master branch? or should we always attach it to a certain version tag? What rule should we follow?
@szha @cjolivier01 @piiswrong

**Comment by marcoabreu** [comment]
I'd vote for using a stable version instead of the latest master

**Comment by zheng-da** [comment]
are you sure this is a MKL problem? it fails in all configurations.

File: tests/cpp/operator/mkldnn.cc [hunk 1]
```diff
@@ -77,4 +77,11 @@ TEST(MKLDNN_UTIL_FUNC, AlignMem) {
   LOG(INFO) << "Skipped for GCC " << __GNUC__ << "." << __GNUC_MINOR__;
 #endif
 }
+
+TEST(MKLDNN_UTIL_FUNC, MemFormat) {
+  // Check whether the number of format is correct.
+  CHECK_EQ(mkldnn_format_last, 56);
```
Comment (cjolivier01) [inline tests/cpp/operator/mkldnn.cc]: Is mkldnn_format_last an enum or constant? If so, you can use static_assert<> somewhere in the code  and it doesn't have to be a unit test.

**Comment by pengzhao-intel** [comment]
@zheng-da will double check. The tests in python2/3 MKLDNN-CPU passed.

**Comment by xinyu-intel** [comment]
I think the failure of this ut test may be related to this old version of mkl-dnn.
https://github.com/apache/incubator-mxnet/blob/5245ef68191a6d47594bf331ec6e20ba6e93ad4c/ci/docker/install/ubuntu_mklml.sh#L24
How to build and test with mkl-dnn in 3rdparty?

**Comment by zheng-da** [comment]
@xinyu-intel I guess you mean mklml?

File: tests/cpp/operator/mkldnn.cc (hunk 1 above)
Comment (TaoLv) [reply tests/cpp/operator/mkldnn.cc]: Yes, it's an enum. This request is from @marcoabreu. Please refer https://github.com/apache/incubator-mxnet/pull/9918#discussion_r171472242 and [jira#98](https://issues.apache.org/jira/browse/MXNET-98) for the background of this. Thanks.

**Comment by xinyu-intel** [comment]
@zheng-da yes, I've made a mistake. It's mklml not mkldnn:)

**Comment by xinyu-intel** [comment]
I have tried the following four tests with seed(1):

First two passed:
exe1 = y1.simple_bind(**mx.cpu()**, x=shape)
exe2 = y2.simple_bind(**mx.cpu()**, x=shape, w=(num_filter, shape[1]//num_group)+kernel, b=(num_filter,))

exe1 = y1.simple_bind(**mx.cpu()**, x=shape)
exe2 = y2.simple_bind(**mx.gpu()**, x=shape, w=(num_filter, shape[1]//num_group)+kernel, b=(num_filter,))

Others failed:

exe1 = y1.simple_bind(**mx.gpu()**, x=shape)
exe2 = y2.simple_bind(**mx.cpu()**, x=shape, w=(num_filter, shape[1]//num_group)+kernel, b=(num_filter,))

(mismatch 94.4444444444%)
 x: array([[[[  11.774242,   10.667873,   37.325356],
         [ -45.697014,   59.5456  ,  -50.37157 ],
         [ -39.387352,  -65.5543  ,   -1.68909 ]]],...
 y: array([[[[  11.774241,   59.5456  ,   -1.689087],
         [  41.73616 ,   95.499115,   16.014626],
         [  12.258306,   22.502499,   45.119247]]],...

exe1 = y1.simple_bind(**mx.gpu()**, x=shape)
exe2 = y2.simple_bind(**mx.gpu()**, x=shape, w=(num_filter, shape[1]//num_group)+kernel, b=(num_filter,))

(mismatch 94.4444444444%)
 x: array([[[[  11.77424 ,   10.667873,   37.32536 ],
         [ -45.697014,   59.5456  ,  -50.37157 ],
         [ -39.387352,  -65.5543  ,   -1.68909 ]]],...
 y: array([[[[  11.774241,   59.545593,   -1.689092],
         [  41.736153,   95.49912 ,   16.014624],
         [  12.258305,   22.5025  ,   45.119247]]],...

It seems that this test cannot pass when using GPU to compute exe1.

**Comment by pengzhao-intel** [comment]
@marcoabreu @cjolivier01  @zheng-da I think the conclusion (based on @xinyu-intel 's analysis) is latest MKL-DNN fixed the problem but the GPU results of `exe1` are not correct.  Could anyone look into GPU side?

@marcoabreu Regarding MKL-DNN, there's release version we can use. But the MXNET development progress is very fast so more new features (or bugfix) are needed. Thus, I think it's OK to select the master branch (based on a commit id). Each CI in MKL-DNN is fully verified and tested.
https://github.com/intel/mkl-dnn/releases

BTW, as we see enabling all test cases would be a great practice to improve the quality.

**Comment by marcoabreu** [comment]
> @marcoabreu Regarding MKL-DNN, there's release version we can use. But the MXNET development progress is very fast so more new features (or bugfix) are needed. Thus, I think it's OK to select the master branch (based on a commit id). Each CI in MKL-DNN is fully verified and tested.
https://github.com/intel/mkl-dnn/releases

I'm indifferent about this one, at least for CI. My only concern is when we make a release with MXNet, we're unable to expect our users to use a (potentially) unstable master commit - usually people prefer to use a stable release. This means that users could run into problems because we're validating against a version of MKLDNN which is not even out yet. We have to consider this fact and have to find a solution - e.g. Intel making more frequent releases of the library, back-porting these fixes or something else along those lines. In the end, we don't want to make a release of MXNet that requires a dependency which is not even out yet. 


> BTW, as we see enabling all test cases would be a great practice to improve the quality.

Definitely! I appreciate efforts in that direction by a lot!


Thanks a lot everybody for all your efforts!

**Comment by zheng-da** [comment]
@xinyu-intel @pengzhao-intel could you describe what is the root of this problem?
It's a little weird that both MKL-DNN and CuDNN have the same bug. Is the bug in the convolution operator? Does the native implementation of MXNet have the bug? Thanks.

**Comment by xinyu-intel** [comment]
@nihui Please help take a look at this pr. The gpu unit test **'depth_wise_conv'** which skipped in #10098 can't pass now. Thank you:)

**Comment by pengzhao-intel** [comment]
ping @nihui

**Comment by nihui** [comment]
@xinyu-intel  hello

I just played with the latest code https://github.com/apache/incubator-mxnet/commit/f3c01d55427bdf16d6478855acbece5009916e10

build with cuda 8.0.61, and without mkl

I uncomment the skip line and remove all testcase except the depth_wise_conv one
and the test passed whatever I changed the binded device to mx.cpu() or mx.gpu()

```
[nihuini@TENCENT64 ~/incubator-mxnet]$ LD_LIBRARY_PATH=/usr/local/cuda/lib64 CUDA_VISIBLE_DEVICES=4,5,6,7 nosetests --verbose --nocapture tests/python/unittest
[INFO] Setting module np/mx/python random seeds, use MXNET_MODULE_SEED=1054985103 to reproduce.
test_operator.test_depthwise_convolution ... ok

----------------------------------------------------------------------
Ran 1 test in 4.753s

OK
```

**Comment by xinyu-intel** [comment]
@nihui Thanks. A bit confused. I just tested on Tesla P100 and got the same error as before.

**Comment by xinyu-intel** [comment]
@nihui Can you please help double check base on the following code:
```
                             dev = default_context()
-                            exe1 = y1.simple_bind(dev, x=shape)
-                            exe2 = y2.simple_bind(mx.cpu(), x=shape, w=(num_filter, shape[1]//num_group)+kernel,
+                            exe1 = y1.simple_bind(mx.gpu(), x=shape)
+                            exe2 = y2.simple_bind(mx.gpu(), x=shape, w=(num_filter, shape[1]//num_group)+kernel,
                                     b=(num_filter,))
```
Thank you!

**Comment by xinyu-intel** [comment]
Thanks, I will retrigger unit test of this pr as soon as #10578 been merged.

**Comment by xinyu-intel** [comment]
All commits have been merged with #10578 , close this pr.

**Pull request closed without merging**